File: python/src/elizaos_plugin_moltbook/lib/intelligence.py
"""
Community Intelligence and Analysis

Analyzes community to find engagement opportunities.

Python port of TypeScript intelligence from plugin-moltbook/typescript/src/lib/intelligence.ts
"""

import json
import logging
from typing import Any, List, Optional
from ..types import CommunityContext, EngagementOpportunity, MoltbookPost, MoltbookProfile
from .templates import get_community_analysis_template

logger = logging.getLogger(__name__)


async def analyze_community(
    posts: List[MoltbookPost],
    character_name: str,
    llm_generate_fn: Any
) -> Optional[CommunityContext]:
    """
    Analyze community to understand trends and find engagement opportunities
    
    Args:
        posts: Recent posts to analyze
        character_name: Agent's name for personalized analysis
        llm_generate_fn: Async function that calls LLM with a prompt
    
    Returns:
        CommunityContext if successful, None otherwise
    """
    
    if not posts:
        logger.warning("No posts to analyze")
        return None
    
    prompt = get_community_analysis_template(
        posts=posts,
        character_name=character_name
    )
    
    try:
        logger.debug(f"Analyzing community ({len(posts)} posts)")
        
        # Get LLM analysis
        response = await llm_generate_fn(prompt)
        
        # Parse JSON response
        try:
            result = json.loads(response)
            
            active_topics = result.get('activeTopics', [])
            vibe = result.get('vibe', 'Unknown')
            what_works = result.get('whatWorks', [])
            engagement_opps_data = result.get('engagementOpportunities', [])
            
            # Parse engagement opportunities
            engagement_opportunities = []
            for opp in engagement_opps_data:
                post_id = opp.get('postId')
                # Find the post
                matching_post = next((p for p in posts if p.get('id') == post_id), None)
                if matching_post:
                    engagement_opportunities.append(
                        EngagementOpportunity(
                            post=matching_post,
                            reason=opp.get('reason', ''),
                            type=opp.get('type', 'comment'),
                            priority=opp.get('priority', 5)
                        )
                    )
            
            # Sort by priority (highest first)
            engagement_opportunities.sort(key=lambda x: x.priority, reverse=True)
            
            # Extract notable community members from posts
            notable_moltys: List[MoltbookProfile] = []
            seen_users = set()
            for post in posts:
                author = post.get('author')
                if author and author.get('username') not in seen_users:
                    notable_moltys.append(author)
                    seen_users.add(author.get('username'))
                if len(notable_moltys) >= 5:
                    break
            
            context = CommunityContext(
                activeTopics=active_topics,
                engagementOpportunities=engagement_opportunities,
                whatWorks=what_works,
                notableMoltys=notable_moltys,
                vibe=vibe,
                analyzedAt=int(__import__('time').time() * 1000)
            )
            
            logger.info(f"Community analyzed: {len(active_topics)} topics, {len(engagement_opportunities)} opportunities")
            return context
            
        except (json.JSONDecodeError, KeyError) as e:
            logger.error(f"Failed to parse community analysis: {e}")
            return None
            
    except Exception as e:
        logger.error(f"Error analyzing community: {e}")
        return None


def find_engagement_opportunities(
    posts: List[MoltbookPost],
    character_interests: List[str]
) -> List[EngagementOpportunity]:
    """
    Simple heuristic-based engagement opportunity finder
    (backup when LLM analysis isn't available)
    
    Args:
        posts: Posts to analyze
        character_interests: Topics the agent is interested in
    
    Returns:
        List of engagement opportunities
    """
    
    opportunities = []
    
    for post in posts:
        title = post.get('title', '').lower()
        content = post.get('content', '').lower()
        score = post.get('score', 0)
        comment_count = post.get('commentCount', 0)
        
        # Calculate priority based on engagement and relevance
        priority = 5
        reason_parts = []
        
        # High engagement = interesting to community
        if score > 10:
            priority += 2
            reason_parts.append("high engagement")
        
        # Few comments = opportunity to add value
        if comment_count < 5:
            priority += 1
            reason_parts.append("open discussion")
        
        # Matches interests
        for interest in character_interests:
            if interest.lower() in title or interest.lower() in content:
                priority += 3
                reason_parts.append(f"relevant to {interest}")
                break
        
        if priority > 5:  # Only include if priority increased
            opportunities.append(
                EngagementOpportunity(
                    post=post,
                    reason=", ".join(reason_parts) if reason_parts else "potential engagement",
                    type='comment',
                    priority=min(priority, 10)
                )
            )
    
    # Sort by priority
    opportunities.sort(key=lambda x: x.priority, reverse=True)
    
    return opportunities[:10]  # Top 10


def format_community_context(context: CommunityContext) -> str:
    """Format community context into human-readable string"""
    
    lines = [
        "## Community Analysis",
        "",
        f"**Vibe**: {context.vibe}",
        "",
        "**Hot Topics**:",
    ]
    
    for topic in context.activeTopics:
        lines.append(f"- {topic}")
    
    lines.extend(["", "**What Works**:"])
    for pattern in context.whatWorks:
        lines.append(f"- {pattern}")
    
    lines.extend(["", f"**Engagement Opportunities**: {len(context.engagementOpportunities)}"])
    for i, opp in enumerate(context.engagementOpportunities[:5], 1):
        post_title = opp.post.get('title', 'Unknown')
        lines.append(f"{i}. [{opp.priority}/10] {post_title[:60]}... - {opp.reason}")
    
    return "\n".join(lines)

File: typescript/src/lib/intelligence.ts
/**
 * Community Intelligence Module
 *
 * Analyzes the Moltbook community to understand what's happening,
 * what topics are hot, and what kinds of content resonate.
 */

import type { IAgentRuntime } from "@elizaos/core";
import { CACHE_TTL_ANALYSIS_MS } from "../constants";
import type {
  CommunityContext,
  EngagementOpportunity,
  MoltbookFeed,
  MoltbookPost,
  MoltbookProfile,
} from "../types";

/**
 * Analyze the community feed to extract context
 */
export function analyzeCommunity(feed: MoltbookFeed, runtime: IAgentRuntime): CommunityContext {
  const posts = feed.posts;

  return {
    activeTopics: extractActiveTopics(posts),
    engagementOpportunities: findEngagementOpportunities(posts, runtime),
    whatWorks: analyzeWhatWorks(posts),
    notableMoltys: findNotableMoltys(posts),
    vibe: assessCommunityVibe(posts),
    analyzedAt: Date.now(),
  };
}

/**
 * Extract trending topics from posts
 */
function extractActiveTopics(posts: MoltbookPost[]): string[] {
  // Word frequency analysis with stopwords filtered
  const stopwords = new Set([
    "the",
    "a",
    "an",
    "is",
    "are",
    "was",
    "were",
    "be",
    "been",
    "being",
    "have",
    "has",
    "had",
    "do",
    "does",
    "did",
    "will",
    "would",
    "could",
    "should",
    "may",
    "might",
    "must",
    "shall",
    "can",
    "need",
    "dare",
    "ought",
    "used",
    "to",
    "of",
    "in",
    "for",
    "on",
    "with",
    "at",
    "by",
    "from",
    "as",
    "into",
    "through",
    "during",
    "before",
    "after",
    "above",
    "below",
    "between",
    "under",
    "again",
    "further",
    "then",
    "once",
    "here",
    "there",
    "when",
    "where",
    "why",
    "how",
    "all",
    "each",
    "few",
    "more",
    "most",
    "other",
    "some",
    "such",
    "no",
    "nor",
    "not",
    "only",
    "own",
    "same",
    "so",
    "than",
    "too",
    "very",
    "just",
    "and",
    "but",
    "if",
    "or",
    "because",
    "until",
    "while",
    "this",
    "that",
    "these",
    "those",
    "am",
    "it",
    "its",
    "they",
    "them",
    "their",
    "what",
    "which",
    "who",
    "whom",
    "i",
    "you",
    "he",
    "she",
    "we",
    "my",
    "your",
    "his",
    "her",
    "our",
    "me",
    "him",
    "us",
    "about",
    "like",
    "get",
    "got",
    "make",
    "made",
    "know",
    "think",
    "see",
    "come",
    "want",
    "look",
    "use",
    "find",
    "give",
    "tell",
    "work",
    "seem",
    "feel",
    "try",
    "leave",
    "call",
    "good",
    "new",
    "first",
    "last",
    "long",
    "great",
    "little",
    "own",
    "old",
    "right",
    "big",
    "high",
    "different",
    "small",
    "large",
    "next",
    "early",
    "young",
    "important",
    "few",
    "public",
    "bad",
    "same",
    "able",
    "im",
    "dont",
    "youre",
    "thats",
    "ive",
    "weve",
    "theyre",
  ]);

  const wordCounts = new Map<string, number>();

  for (const post of posts) {
    // Combine title and content
    const text = `${post.title} ${post.content}`.toLowerCase();

    // Extract words (alphanumeric only)
    const words = text.match(/\b[a-z][a-z0-9]{2,}\b/g) || [];

    for (const word of words) {
      if (!stopwords.has(word) && word.length > 2) {
        wordCounts.set(word, (wordCounts.get(word) || 0) + 1);
      }
    }
  }

  // Sort by frequency and take top topics
  const sortedWords = [...wordCounts.entries()]
    .filter(([_, count]) => count >= 2) // Appears in at least 2 posts
    .sort((a, b) => b[1] - a[1])
    .slice(0, 10)
    .map(([word]) => word);

  return sortedWords;
}

/**
 * Find posts worth engaging with
 */
function findEngagementOpportunities(
  posts: MoltbookPost[],
  runtime: IAgentRuntime
): EngagementOpportunity[] {
  const opportunities: EngagementOpportunity[] = [];
  const characterName = runtime.character.name?.toLowerCase() || "";
  const characterTopics = extractCharacterTopics(runtime);

  for (const post of posts) {
    const postText = `${post.title} ${post.content}`.toLowerCase();

    // Priority 1: Mentions of the agent or related topics
    if (
      postText.includes(characterName) ||
      postText.includes("elizaos") ||
      postText.includes("eliza")
    ) {
      opportunities.push({
        post,
        reason: "Mentions relevant topic - good opportunity to add perspective",
        type: "comment",
        priority: 10,
      });
      continue;
    }

    // Priority 2: Posts matching character interests
    const matchingTopics = characterTopics.filter((topic) =>
      postText.includes(topic.toLowerCase())
    );
    if (matchingTopics.length > 0) {
      opportunities.push({
        post,
        reason: `Relates to character interests: ${matchingTopics.join(", ")}`,
        type: "comment",
        priority: 8,
      });
      continue;
    }

    // Priority 3: Posts asking questions
    if (post.title.includes("?") || post.content.includes("?")) {
      opportunities.push({
        post,
        reason: "Post is asking a question - could provide helpful answer",
        type: "comment",
        priority: 6,
      });
      continue;
    }

    // Priority 4: High-engagement posts worth upvoting
    if (post.score > 10 && post.commentCount > 5) {
      opportunities.push({
        post,
        reason: "Popular post worth acknowledging",
        type: "upvote",
        priority: 3,
      });
    }

    // Priority 5: New users worth following
    if (post.author.postCount < 5 && post.score > 0) {
      opportunities.push({
        post,
        reason: "New community member with quality content",
        type: "follow",
        priority: 2,
      });
    }
  }

  // Sort by priority and return top opportunities
  return opportunities.sort((a, b) => b.priority - a.priority).slice(0, 10);
}

/**
 * Extract topics from character configuration
 */
function extractCharacterTopics(runtime: IAgentRuntime): string[] {
  const topics: string[] = [];

  // From character bio
  if (runtime.character.bio) {
    const bioText = Array.isArray(runtime.character.bio)
      ? runtime.character.bio.join(" ")
      : runtime.character.bio;
    // Simple keyword extraction from bio
    const keywords = bioText.match(/\b[A-Za-z][a-z]{3,}\b/g) || [];
    topics.push(...keywords.slice(0, 10));
  }

  // From character topics if defined
  if (runtime.character.topics) {
    topics.push(...runtime.character.topics);
  }

  // From character adjectives (personality traits)
  if (runtime.character.adjectives) {
    topics.push(...runtime.character.adjectives);
  }

  return [...new Set(topics)]; // Dedupe
}

/**
 * Analyze what content patterns work well
 */
function analyzeWhatWorks(posts: MoltbookPost[]): string[] {
  const patterns: string[] = [];

  // Analyze high-scoring posts
  const highScorePosts = posts.filter((p) => p.score > 5);

  if (highScorePosts.length > 0) {
    // Check average title length
    const avgTitleLength =
      highScorePosts.reduce((sum, p) => sum + p.title.length, 0) / highScorePosts.length;
    if (avgTitleLength < 50) {
      patterns.push("Concise titles (under 50 characters) perform well");
    } else if (avgTitleLength > 100) {
      patterns.push("Descriptive titles work in this community");
    }

    // Check for question posts
    const questionPosts = highScorePosts.filter((p) => p.title.includes("?"));
    if (questionPosts.length > highScorePosts.length * 0.3) {
      patterns.push("Questions engage the community");
    }

    // Check content length
    const avgContentLength =
      highScorePosts.reduce((sum, p) => sum + p.content.length, 0) / highScorePosts.length;
    if (avgContentLength < 500) {
      patterns.push("Short, punchy posts get engagement");
    } else if (avgContentLength > 1500) {
      patterns.push("In-depth content is valued");
    }
  }

  // Fallback patterns
  if (patterns.length === 0) {
    patterns.push("Share unique perspectives and experiences");
    patterns.push("Ask thought-provoking questions");
    patterns.push("Provide value in every post");
  }

  return patterns;
}

/**
 * Find notable community members
 */
function findNotableMoltys(posts: MoltbookPost[]): MoltbookProfile[] {
  const authorScores = new Map<
    string,
    { profile: MoltbookProfile; totalScore: number; postCount: number }
  >();

  for (const post of posts) {
    const existing = authorScores.get(post.authorId);
    if (existing) {
      existing.totalScore += post.score;
      existing.postCount += 1;
    } else {
      authorScores.set(post.authorId, {
        profile: post.author,
        totalScore: post.score,
        postCount: 1,
      });
    }
  }

  // Sort by engagement and return top authors
  return [...authorScores.values()]
    .filter((a) => a.postCount >= 2 || a.totalScore > 10)
    .sort((a, b) => b.totalScore - a.totalScore)
    .slice(0, 5)
    .map((a) => a.profile);
}

/**
 * Assess the overall community vibe
 */
function assessCommunityVibe(posts: MoltbookPost[]): string {
  if (posts.length === 0) {
    return "quiet - not much activity right now";
  }

  const avgScore = posts.reduce((sum, p) => sum + p.score, 0) / posts.length;
  const avgComments = posts.reduce((sum, p) => sum + p.commentCount, 0) / posts.length;

  // Check for recent activity
  const now = Date.now();
  const recentPosts = posts.filter((p) => {
    const postTime = new Date(p.createdAt).getTime();
    return now - postTime < 24 * 60 * 60 * 1000; // Last 24 hours
  });

  let vibe = "";

  if (recentPosts.length > posts.length * 0.5) {
    vibe = "active";
  } else {
    vibe = "steady";
  }

  if (avgScore > 10) {
    vibe += ", supportive";
  } else if (avgScore < 2) {
    vibe += ", discerning";
  }

  if (avgComments > 5) {
    vibe += ", conversational";
  }

  return vibe || "engaged";
}

/**
 * Check if cached analysis is still fresh
 */
export function isAnalysisFresh(
  context: CommunityContext | undefined,
  maxAgeMs: number = CACHE_TTL_ANALYSIS_MS
): boolean {
  if (!context) return false;
  return Date.now() - context.analyzedAt < maxAgeMs;
}

/**
 * Format community context for inclusion in prompts
 */
export function formatContextForPrompt(context: CommunityContext): string {
  const lines: string[] = [];

  lines.push("## Current Moltbook Community Context");
  lines.push("");

  if (context.activeTopics.length > 0) {
    lines.push(`**Hot Topics:** ${context.activeTopics.slice(0, 5).join(", ")}`);
  }

  lines.push(`**Community Vibe:** ${context.vibe}`);

  if (context.whatWorks.length > 0) {
    lines.push("");
    lines.push("**What Works Here:**");
    for (const pattern of context.whatWorks.slice(0, 3)) {
      lines.push(`- ${pattern}`);
    }
  }

  if (context.engagementOpportunities.length > 0) {
    lines.push("");
    lines.push("**Engagement Opportunities:**");
    for (const opp of context.engagementOpportunities.slice(0, 3)) {
      lines.push(`- "${opp.post.title}" - ${opp.reason}`);
    }
  }

  if (context.notableMoltys.length > 0) {
    lines.push("");
    lines.push(
      `**Notable Community Members:** ${context.notableMoltys.map((m) => m.username).join(", ")}`
    );
  }

  return lines.join("\n");
}
